### PR TITLE
EntityTypes

### PR DESCRIPTION
**Pull request by TheGlitch76**
Includes a new module, intended to hold vanilla bugfixes and other odd behavior changes to Vanilla behavior that Forge makes and mods rely on. This is better than keeping the bugfixes in their own modules because it encourages developers to make their fixes in ways that will not conflict with other mods (i.e. Patchwork modules). This will also aid in updating because we can just go down the list of bugs in one place, instead of searching the entire Patchwork codebase.

### CODE REVIEW

File: patchwork-extensions-shearing/src/main/java/com/patchworkmc/impl/extension/shearing/Shearables.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
  */
 
-package com.patchworkmc.impl.extensions.shearing;
+package com.patchworkmc.impl.extension.shearing;
```
Comment (TheGlitch76) [inline patchwork-extensions-shearing/src/main/java/com/patchworkmc/impl/extension/shearing/Shearables.java]: This is an unrelated fix, this package name didn't match the ones used everywhere else.

Comment (coderbot16) [reply patchwork-extensions-shearing/src/main/java/com/patchworkmc/impl/extension/shearing/Shearables.java]: No? `com.patchworkmc.mixin.extensions.shearing` is what all of the shearing mixins use, and MinecraftForge uses `net.minecraftforge.common.extensions`. The only exception is MixinStatusEffect. If you want to rename all of the other patchwork packages, fine, but this is only halfway. I think this refactor makes sense for another PR. I'm still split if `extension` or `extensions` makes more sense for the package name: Forge uses `extensions`, and most of patchwork uses `extensions`.

File: patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityCategory;
+import net.minecraft.entity.EntityDimensions;
+import net.minecraft.entity.EntityType;
+
+import com.patchworkmc.impl.patches.bugfixes.EntityTypeBuilderShim;
+
+@Mixin(EntityType.Builder.class)
+@Implements(@Interface(iface = EntityTypeBuilderShim.class, prefix = "patchwork$"))
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java]: Is the interface needed here? I don't think this is ever accessed externally.

File: patchwork-vanilla-patches/src/README.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# patchwork-vanilla-patches
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/README.md]: I generally put a blank line between headers and content.

File: patchwork-vanilla-patches/src/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+A collection of bugfixes and other changes to vanilla behavior that Forge makes.
+
+Bugfixes should be put here instead of other modules whenever possible, in order to force those fixes to be done in a way
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/README.md]: Generally I use softwrap for markdown, and don't manually wrap lines.

File: patchwork-vanilla-patches/src/README.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+that it doesn't break other mods.
+
+Current changes:
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/README.md]:
```suggestion
## Current changes

```

File: patchwork-extensions/src/main/resources/patchwork-extensions.mixins.json [hunk 6]
```diff
@@ -3,7 +3,10 @@
   "package": "com.patchworkmc.mixin.extension",
   "compatibilityLevel": "JAVA_8",
   "mixins": [
-    "MixinStatusEffect"
+    "MixinEntityType",
+    "MixinEntityTypeBuilder",
+    "MixinStatusEffect",
+    "MixinEntityTypeBuilder"
```
Comment (coderbot16) [inline patchwork-extensions/src/main/resources/patchwork-extensions.mixins.json]: Duplicate entry?

File: patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeExtensions.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package com.patchworkmc.impl.extension;
+
+import net.minecraft.entity.Entity;
+
+public interface PatchworkEntityTypeExtensions<T extends Entity> {
+	// Patchwork: Forge does this through patching the constructor instead
```
Comment (coderbot16) [inline patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeExtensions.java]: Make this a javadoc of the interface?

File: patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java (hunk 2 above)
Comment (TheGlitch76) [reply patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java]: You need the shim for the proxy overwrite thing

Comment (coderbot16) [reply patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java]: Is it 100% necessary? Have you tested without it?

File: patchwork-vanilla-patches/src/README.md (hunk 4 above)
Comment (TheGlitch76) [reply patchwork-vanilla-patches/src/README.md]: 156 columns seems awful long for editing in IDEA/etc, which lacks soft wrap.

Comment (TheGlitch76) [reply patchwork-vanilla-patches/src/README.md]: Either way, a one line wrap does not affect the actual displayed file.

File: patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+package com.patchworkmc.impl.extension;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityType;
+
+/**
+ * Forge does this through patching the constructor instead, we just add methods with mixins instead.
+ */
+public interface PatchworkEntityTypeBuilderExtensions<T extends Entity> {
```
Comment (shedaniel) [inline patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java]: One question? why is this class required? You don't need an interface to add methods iirc.

File: patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	private boolean field_19424;
+
+	@Shadow
+	private EntityDimensions size;
+
+	/**
+	 * Fixes MC-170128: Cannot build an EntityType without a datafixer due to an IllegalArgumentException.
+	 */
+	@Intrinsic(displace = true)
+	public EntityType<T> patchwork$build(String id) {
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java]: After some discussion in #mixin, I don't think that this will work in production, since it doesn't get remapped. We'll need another approach perhaps.

File: patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java (hunk 8 above)
Comment (TheGlitch76) [reply patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java]: Maybe not (i'm not sure from the docs), but not including this makes it impossible to reference our added methods with duck typing.

**Comment by florensie** [comment]
The icon should be in `resources/assets/patchwork-vanilla-patches/`. You forgot the assets directory.

File: patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java (hunk 8 above)
Comment (coderbot16) [reply patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java]: On one hand we don't need this because we don't use this directly, on the other other mods may want to compile against this. I'm still uncertain whether it's necessary.

File: patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+	EntityType.Builder<T> setUpdateInterval(int interval);
+
+	EntityType.Builder<T> setTrackingRange(int range);
+
+	EntityType.Builder<T> setShouldReceiveVelocityUpdates(boolean value);
+
+	// TODO: setCustomClientFactory
```
Comment (coderbot16) [inline patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java]: Note that setCustomClientFactory will be handled by the networking module.

File: patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeExtensions.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * Forge does this through patching the constructor instead, we just add methods with mixins instead.
+ */
+public interface PatchworkEntityTypeExtensions<T extends Entity> {
+	void setUpdateInterval(int interval);
+
+	void setTrackingRange(int range);
+
+	void setShouldReceiveVelocityUpdates(boolean value);
+
+	// TODO: setCustomClientFactory
```
Comment (coderbot16) [inline patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeExtensions.java]: Same here

File: patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+import net.minecraft.entity.EntityType;
+
+@Mixin(EntityType.Builder.class)
+public abstract class MixinEntityTypeBuilder<T extends Entity> {
+	@Unique
+	private static final Logger LOGGER = LogManager.getLogger();
+
+	@Shadow
+	@Final
+	private EntityType.EntityFactory<T> factory;
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java]: All of these Shadows appear to be unused....

File: patchwork-vanilla-patches/src/main/resources/fabric.mod.json [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  "contact": {
+    "issues": "https://github.com/PatchworkMC/patchwork-api/issues",
+    "sources": "https://github.com/PatchworkMC/patchwork-api"
+  },
+  "authors": [
+    "PatchworkMC"
+  ],
+  "depends": {
+    "fabricloader": ">=0.6.2",
+    "patchwork-fml": "*"
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/main/resources/fabric.mod.json]: Does it depend on patchwork-fml? It's not in the build gradle.

Comment (TheGlitch76) [reply patchwork-vanilla-patches/src/main/resources/fabric.mod.json]: it's an oversight, will remove it

File: patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java (hunk 8 above)
Comment (TheGlitch76) [reply patchwork-extensions/src/main/java/com/patchworkmc/impl/extension/PatchworkEntityTypeBuilderExtensions.java]: If we have patchwork-patcher patching source code this is required.

File: patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+@Mixin(EntityType.Builder.class)
+public abstract class MixinEntityTypeBuilder<T extends Entity> {
+	@Unique
+	private static final Logger LOGGER = LogManager.getLogger();
+
+	/**
+	 * Fixes MC-170128: Cannot build an EntityType without a datafixer due to an IllegalArgumentException.
+	 */
+	@Redirect(method = "build", at = @At(value = "INVOKE", target = "Lcom/mojang/datafixers/schemas/Schema;getChoiceType(Lcom/mojang/datafixers/DSL$TypeReference;Ljava/lang/String;)Lcom/mojang/datafixers/types/Type;"), remap = false)
```
Comment (coderbot16) [inline patchwork-vanilla-patches/src/main/java/com/patchworkmc/mixin/patches/bugfixes/MixinEntityTypeBuilder.java]: `build` should not get remapped.
```suggestion
	@Redirect(method = "build", at = @At(value = "INVOKE", target = "Lcom/mojang/datafixers/schemas/Schema;getChoiceType(Lcom/mojang/datafixers/DSL$TypeReference;Ljava/lang/String;)Lcom/mojang/datafixers/types/Type;", remap = false))
```
